internal/impl/git/input_config.go

```diff
@@ -0,0 +1,342 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package git
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/bmatcuk/doublestar/v4"
+	"github.com/redpanda-data/benthos/v4/public/service"
+)
+
+// gitInputConfig returns the configuration specification for the Git input plugin.
+func gitInputConfig() *service.ConfigSpec {
+	desc := `
+The git input clones the specified repository (or pulls updates if already cloned) and reads 
+the content of the specified file. It periodically polls the repository for new commits and emits 
+a message when changes are detected.
+
+== Metadata
+
+This input adds the following metadata fields to each message:
+
+- git_file_path
+- git_file_size
+- git_file_mode
+- git_file_modified
+- git_commit
+- git_mime_type
+- git_is_binary
+- git_encoding (present if the file was base64 encoded)
+- git_deleted (only present if the file was deleted)
+
+You can access these metadata fields using function interpolation.`
+
+	return service.NewConfigSpec().
+		Beta().
+		Categories("Services").
+		Version("4.51.0").
+		Summary(`A Git input that clones (or pulls) a repository and reads the repository contents.`).
+		Description(desc).
+		Fields(
+			// General git cloning & polling settings
+			service.NewStringField("repository_url").
+				Description("The URL of the Git repository to clone.").
+				Example("https://github.com/username/repo.git"),
+			service.NewStringField("branch").
+				Description("The branch to check out.").
+				Default("main"),
+			service.NewDurationField("poll_interval").
+				Description("Duration between polling attempts").
+				Default("10s").
+				Example("10s"),
+			service.NewStringListField("include_patterns").
+				Description("A list of file patterns to include (e.g., '**/*.md', 'configs/*.yaml'). If empty, all files will be included. "+
+					"Supports glob patterns: *, /**/, ?, and character ranges [a-z]. Any character with a special meaning can be escaped with a backslash.").
+				Default([]any{}).
+				Optional(),
+			service.NewStringListField("exclude_patterns").
+				Description("A list of file patterns to exclude (e.g., '.git/**', '**/*.png'). These patterns take precedence over include_patterns. "+
+					"Supports glob patterns: *, /**/, ?, and character ranges [a-z]. Any character with a special meaning can be escaped with a backslash.").
+				Default([]any{}).
+				Optional(),
+			service.NewBoolField("recursive").
+				Description("Whether to recursively read files in subdirectories.").
+				Default(true),
```

Comment:
Personally I think we should remove this option and never emit directories but instead emit full files, this recursive behavior can be changed with the `include_patterns`.